README.md
Django Opensearch DSL
=====================

[![PyPI Version](https://badge.fury.io/py/django-opensearch-dsl.svg)](https://badge.fury.io/py/django-opensearch-dsl)
[![Documentation Status](https://readthedocs.org/projects/django-opensearch-dsl/badge/?version=latest)](https://django-opensearch-dsl.readthedocs.io/en/latest/?badge=latest)
![Tests](https://github.com/Codoc-os/django-opensearch-dsl/workflows/Tests/badge.svg)
[![Python 3.7+](https://img.shields.io/badge/Python-3.7+-brightgreen.svg)](#)
[![Django 3.2+](https://img.shields.io/badge/Django-3.2+-brightgreen.svg)](#)
[![OpenSearch 1.3+, 2.7+](https://img.shields.io/badge/OpenSearch-1.3+-brightgreen.svg)](#)
[![License Apache 2](https://img.shields.io/badge/license-Apache%202-brightgreen.svg)](https://github.com/Codoc-os/django-opensearch-dsl/blob/master/LICENSE)
[![codecov](https://codecov.io/gh/Codoc-os/django-opensearch-dsl/branch/master/graph/badge.svg)](https://codecov.io/gh/Codoc-os/django-opensearch-dsl)
[![CodeFactor](https://www.codefactor.io/repository/github/Codoc-os/django-opensearch-dsl/badge)](https://www.codefactor.io/repository/github/Codoc-os/django-opensearch-dsl)

**Django Opensearch DSL** is a package that allows the indexing of Django models in opensearch. It is built as a thin
wrapper around [`opensearch-py`](https://github.com/opensearch-project/opensearch-py)
so you can use all the features developed by the `opensearch-py` team.

You can view the full documentation
at [https://django-opensearch-dsl.readthedocs.io](https://django-opensearch-dsl.readthedocs.io/en/latest/).

## Features

- Based on [`opensearch-py`](https://github.com/opensearch-project/opensearch-py) so you can make queries with
  the [`Search`](https://elasticsearch-dsl.readthedocs.io/en/latest/search_dsl.html#the-search-object)
  object.
- Management commands for creating, deleting, and populating indices and documents.
- Opensearch auto mapping from Django models fields.
- Complex field type support (`ObjectField`, `NestedField`).
- Index fast using `parallel` indexing.
- Zero-downtime index migration for when the document mapping changes.

## Requirements

`django-opensearch-dsl` only support the supported version of each dependency (mainstream & lts).

* `Python` ([supported versions](https://devguide.python.org/versions/))
* `Django` ([supported versions](https://www.djangoproject.com/download/#supported-versions))
* `opensearch-py` ([compatibility matrix](https://github.com/opensearch-project/opensearch-py/blob/main/COMPATIBILITY.md))

## Installation and Configuration

The easiest way to install `django-opensearch-dsl` is through `pip`:

* `pip install django-opensearch-dsl`

Then add `django_opensearch_dsl` to your `INSTALLED_APPS` settings.

You must then define `OPENSEARCH_DSL` in your Django settings.

For example:

```python
OPENSEARCH_DSL = {
    'default': {
        'hosts': 'localhost:9200'
    },
    'secure': {
        'hosts': [{"scheme": "https", "host": "192.30.255.112", "port": 9201}],
        'http_auth': ("admin", "password"),
        'timeout': 120,
    },
}
```

`OPENSEARCH_DSL` is then passed
to [`opensearchpy.connection.connections.configure`](http://elasticsearch-dsl.readthedocs.io/en/stable/configuration.html#multiple-clusters)
.

## Create Document Classes

To index instances of the following model :

```python
# models.py

class Car(models.Model):
    name = models.CharField()
    color = models.CharField()
    description = models.TextField()
    type = models.IntegerField(choices=[
        (1, "Sedan"),
        (2, "Truck"),
        (4, "SUV"),
    ])
```

First create a subclass of [`django_opensearch_dsl.Document`](document.md) containing the subclasses `Index`
(which define the index' settings) and `Django` (which contains settings related to your django `Model`). Finally,
register the class using `registry.register_document()` decorator.

It is required to define `Document` classes inside a file named `documents.py` in your apps' directory.

```python
# documents.py

from django_opensearch_dsl import Document
from django_opensearch_dsl.registries import registry
from .models import Car


@registry.register_document
class CarDocument(Document):
    class Index:
        name = 'cars'  # Name of the Opensearch index
        settings = {  # See Opensearch Indices API reference for available settings
            'number_of_shards': 1,
            'number_of_replicas': 0
        }
        # Configure how the index should be refreshed after an update.
        # See Opensearch documentation for supported options.
        # This per-Document setting overrides settings.OPENSEARCH_DSL_AUTO_REFRESH.
        auto_refresh = False

    class Django:
        model = Car  # The model associated with this Document        
        fields = [  # The fields of the model you want to be indexed in Opensearch
            'name',
            'color',
            'description',
            'type',
        ]
        # Paginate the django queryset used to populate the index with the specified size
        # This per-Document setting overrides settings.OPENSEARCH_DSL_QUERYSET_PAGINATION.
        queryset_pagination = 5000
```


docs/CHANGELOG.md
# Changelog

## X.Y.Z (not released yet)

* New feature: zero-downtime document migration

### 0.5.1 (2023-05-18)

* Change references from `opensearch-dsl-py` to `opensearch-py`.  
  This follow the deprecation notice on the
  [`opensearch-dsl-py`](https://github.com/opensearch-project/opensearch-dsl-py) project. Its features are now directly
  included in `opensearch-py`.  
  ([#33](https://github.com/Codoc-os/django-opensearch-dsl/issues/33), Contributed by [Jacob Kausler](https://github.com/jakekausler)).


## 0.5.0 (2022-11-19)

* `get_indexing_queryset()` now order unordered QuerySet by their PK.
  ([#29](https://github.com/Codoc-os/django-opensearch-dsl/issues/29), Contributed by [Cédric Raud](https://github.com/cedricraud)).
* `keep_order` argument of `django_opensearch_dsl.search.Search.to_queryset` now default to `True`
  to be in line with the documentation ([#27](https://github.com/Codoc-os/django-opensearch-dsl/issues/27)).

### 0.4.1 (2022-08-16)

* `Document.update()` method now take an optional `using` argument allowing to specify an alternate
  OpenSearch connection defined in `OPENSEARCH_DSL`.
* Fix related document automatic indexation and deletion
  (Contributed by [Colin Seifer](https://github.com/Colin-Seifer)).
* Add `pre-delete` back into `BaseSignalProcessor.handle_m2m_changed()` to  properly update the
  index on M2M interactions (Contributed by [Colin Seifer](https://github.com/Colin-Seifer)).

## 0.4.0 (2022-08-04)

* Add support for related models. See [Document Classes](document.md#django-subclass) and
  [Document Field Reference](fields.md#using-prepare_field_with_related) for more information
  (Contributed by [Colin Seifer](https://github.com/Colin-Seifer)).
* `django-opensearch-dsl` now only tests supported version of Python and Django (mainstream and LTS). 
  This choice is made to:
    * Speed up development.
    * Speed up tests.
    * Reduce actions on github.
    * Encourage people to update their stack to supported (thus safer) versions.
* Drop support for Python 3.6.
* Drop support for Django 2.1, 2.2, 3.1.
* Now supports Django 4.1.
* Now supports `opensearch-dsl>=2.0 <3.00`.

## 0.3.0 (2022-06-22)

* Fixes internal links in documentation.
* Remove the need to declare a `TESTING` boolean in `settings.py`.

## 0.2.0 (2022-01-13)

* Restore auto-sync feature (still undocumented - Contributed by [David Guillot](https://github.com/David-Guillot))
* Add support to Django 4.0 (Contributed by [David Guillot](https://github.com/David-Guillot))
* Remove some python2 leftovers (Contributed by [David Guillot](https://github.com/David-Guillot))

## 0.1.2 (2021-12-14)

* Fixed 'Search.validate()'

## 0.1.0 (2021-12-11)

* Migrated to **Opensearch**
* Drop some feature such as auto-syncing signals and related models
* Replace `search_index` management command with `opensearch`.

## 0.1.0 (2021-12-11)

* Migrated to **Opensearch**
* Drop some feature such as auto-syncing signals and related models
* Replace `search_index` management command with `opensearch`.

## Before fork from `django-elasticsearch-dsl`

### 7.1.4 (2020-07-05)

* Configure Elasticsearch _id dynamically from document (#272)
* Use chain.from_iterable in for performance improvement (#278)
* Handle case where SimpleLazyObject being treated as an Iterable (#255)
* Camelcase default value in management command (#254)
* Various updates and fixup in docs (#250, #276)
* Start testing against Python 3.8 (#266)

### 7.1.1 (2019-12-26)

* Adding detailed documentation and published to Read The Docs #222
* Resolve name resolution while delete, create index (#228)
* Added support for Django 3.0. (#230)
* Removing old Elasticsearc compatibility (#219)
* Drop StringField in favor of TextField.

### 7.1.0 (2019-10-29)

* Support for Django `DecimalField` #141
* Indexing speedup by using `parallel` indexing. #213. Now you can pass `--parallel` or set `ELASTICSEARCH_DSL_PARALLEL`
  in your settings to get indexing speed boost while indexing through management command.
* Fixing name resolution in management command #206
* Small documentation fixes. #196

### 7.0.0 (2019-08-11)

* Support Elasticsearch 7.0 (See PR #176)
* Added order by to paginate queryset properly (See PR #153)
* Remove `standard` token filter from `README.md` and test files
* Various documentation fixes

### 6.4.2 (2019-07-26)

* Fix document importing path
* Update readme

### 6.4.1 (2019-06-14)

* The `DocType` import has changed to `Document`

### 6.4.0 (2019-06-01)

* Support elasticsearch-dsl>6.3.0
* Class `Meta` has changed to class `Django` (See PR #136)
* Add `register_document` decorator to register a document (See PR #136)
* Additional Bug fixing and others

### 0.5.1 (2018-11-07)

* Limit elastsearch-dsl to supported versions

### 0.5.0 (2018-04-22)

* Add Support for Elasticsearch 6 thanks to HansAdema

### Breaking Change:

* Django string fields now point to ES text field by default.
* Nothing should change for ES 2.X but if you are using ES 5.X, you may need to rebuild and/or update some of your
  documents.

### 0.4.5 (2018-04-22)

* Fix prepare with related models when deleted (See PR #99)
* Fix unwanted calls to get_instances_from_related
* Fix for empty ArrayField (CBinyenya)
* Fix nested OneToOneField when related object doesn't exist (CBinyenya)
* Update elasticsearch-dsl minimal version

### 0.4.4 (2017-12-13)

* Fix to_queryset with es 5.0/5.1

### 0.4.3 (2017-12-12)

* Fix syncing of related objects when deleted
* Add django 2.0 support

### 0.4.2 (2017-11-27)

* Convert lazy string to string before serialization
* Readme update (arielpontes)

### 0.4.1 (2017-10-17)

* Update example app with get_instances_from_related
* Typo/grammar fixes

### 0.4.0 (2017-10-07)

* Add a method on the Search class to return a django queryset from an es result
* Add a queryset_pagination option to DocType.Meta for allow the pagination of big django querysets during the index
  populating

tests/tests/test_search.py
from django.test import TestCase
from opensearchpy.helpers.query import Query
from opensearchpy.connection.connections import get_connection

from django_dummy_app.documents import CountryDocument, ContinentDocument
from django_dummy_app.models import Country, Continent


class DocumentTestCase(TestCase):
    fixtures = ["tests/django_dummy_app/geography_data.json"]

    def setUp(self) -> None:
        get_connection().indices.delete("_all")

    def test_search_country(self):
        CountryDocument.init()

        CountryDocument().update(CountryDocument().get_indexing_queryset(), "index", refresh=True)
        self.assertEqual(
            set(CountryDocument.search().query("term", **{"continent.name": "Europe"}).extra(size=300).to_queryset()),
            set(Country.objects.filter(continent__name="Europe")),
        )

    def test_search_country_cache(self):
        CountryDocument.init()

        CountryDocument().update(CountryDocument().get_indexing_queryset(), "index", refresh=True)
        search = CountryDocument.search().query("term", **{"continent.name": "Europe"}).extra(size=300)
        search.execute()
        self.assertEqual(
            set(search.to_queryset(keep_order=True)), set(Country.objects.filter(continent__name="Europe"))
        )

    def test_search_country_keep_order(self):
        CountryDocument.init()

        CountryDocument().update(CountryDocument().get_indexing_queryset(), "index", refresh=True)
        search = CountryDocument.search().query("term", **{"continent.name": "Europe"}).extra(size=300)
        self.assertEqual(
            set(search.to_queryset(keep_order=True)), set(Country.objects.filter(continent__name="Europe"))
        )

    def test_search_country_refresh_default_to_document(self):
        CountryDocument.init()

        CountryDocument().update(CountryDocument().get_indexing_queryset(), "index", refresh=True)
        self.assertEqual(
            set(CountryDocument.search().query("term", **{"continent.name": "Europe"}).extra(size=300).to_queryset()),
            set(Country.objects.filter(continent__name="Europe")),
        )

    def test_search_country_refresh_default_to_settings(self):
        ContinentDocument.init()

        ContinentDocument().update(ContinentDocument().get_indexing_queryset(), "index", refresh=True)
        search = ContinentDocument.search().query(
            "nested", path="countries", query=Q("term", **{"countries.name": "France"})
        )
        self.assertEqual(set(search.to_queryset()), set(Continent.objects.filter(countries__name="France")))

    def test_update_instance(self):
        ContinentDocument.init()

        ContinentDocument().update(Continent.objects.get(countries__name="France"), "index", refresh=True)
        search = ContinentDocument.search().query(
            "nested", path="countries", query=Q("term", **{"countries.name": "France"})
        )
        self.assertEqual(set(search.to_queryset()), set(Continent.objects.filter(countries__name="France")))


django_opensearch_dsl/management/types.py
import datetime
from typing import Union, List

from dateutil.parser import isoparse
from django.conf import settings
from django.utils.timezone import make_aware, is_aware

Nothing = type(Ellipsis)
Values = Union[None, int, float, datetime.datetime, str, List["Values"]]


def bool_parser(value: str) -> Union[Nothing, bool]:
    """Try to parse the given value as a boolean."""
    if value.lower() in ("true", "false"):
        return value.lower() == "true"
    return ...

def datetime_parser(value: str) -> Union[Nothing, datetime.datetime]:
    """Try to parse the given value as a ISO 8601's datetime."""
    try:
        date = isoparse(value)
        if not is_aware(date):
            date = make_aware(isoparse(value))
        return date
    except ValueError as e:
        return ...


def int_parser(value: str) -> Union[Nothing, int]:
    """Try to parse the given value as an integer."""
    try:
        return int(value)
    except ValueError:
        return ...


def float_parser(value: str) -> Union[Nothing, float]:
    """Try to parse the given value as a float."""
    try:
        return float(value)
    except ValueError:
        return ...


def none_parser(value: str) -> Union[Nothing, None]:
    """Try to parse the given value as a float."""
    return None if value == "" else ...


def list_parser(value: str) -> Values:
    """Try to parse the given value as a ISO 8601's datetime."""
    if "," in value:
        return [parse(v.strip()) for v in value.split(",")]
    return ...


def parse(value: str) -> Values:
    """Try to coerce a string into a different type.

    The order in which the parsers are called matters as some type might include
    other, e.g. `float_parser()` would match an `int` as a `float` if called
    before `int_parser()`.

    If no parser was able to parse the value, it is returned as a string.
    """
    parsers = getattr(
        settings, "OPENSEARCH_DSL_VALUE_PARSERS", [none_parser, bool_parser, int_parser, float_parser, datetime_parser, list_parser]
    )
    for parser in parsers:
        v = parser(value)
        if v != ...:  # noqa
            return v
    return value


django_opensearch_dsl/management/commands/opensearch.py
import argparse
import functools
import operator
import os
import sys
from argparse import ArgumentParser
from collections import defaultdict
from datetime import datetime
from typing import Any, Callable

import opensearchpy
from django.conf import settings
from django.core.exceptions import FieldError
from django.core.management import BaseCommand
from django.core.management.base import OutputWrapper
from django.db.models import Q

from django_opensearch_dsl.registries import registry
from ..enums import OpensearchAction
from ..types import parse
from ...documents import Document


class Command(BaseCommand):
    """Manage indices and documents."""

    help = (
        "Allow to create and delete indices, as well as indexing, updating or deleting specific "
        "documents from specific indices.\n"
    )

    def __init__(self, *args, **kwargs):  # noqa
        super(Command, self).__init__()
        self.usage = None

    def db_filter(self, parser: ArgumentParser) -> Callable[[str], Any]:
        """Return a function to parse the filters."""

        def wrap(value):  # pragma: no cover
            try:
                lookup, v = value.split("=")
                v = parse(v)
            except ValueError:
                sys.stderr.write(parser._subparsers._group_actions[0].choices["document"].format_usage())  # noqa
                sys.stderr.write(
                    f"manage.py index: error: invalid filter: '{value}' (filter must be formatted as "
                    f"'[Field Lookups]=[value]')\n",
                )
                exit(1)
            return lookup, v  # noqa

        return wrap

    def __list_index(self, **options):  # noqa pragma: no cover
        """List all known index and indicate whether they are created or not."""
        indices = registry.get_indices()
        result = defaultdict(list)
        for index in indices:
            document = index._doc_types[0]
            module = document.__module__.split(".")[-2]  # noqa
            exists = index.exists()
            checkbox = f"[{'X' if exists else ' '}]"
            document_indices = document.get_all_indices()
            if document_indices:
                details = ". Following indices exist:"
                for document_index in document_indices:
                    is_active = ""
                    if document_index.exists_alias(name=index._name):
                        is_active = self.style.SUCCESS("Active")
                    count = f" ({document_index.search().count()} documents)"
                    details += f"\n    - {document_index._name} {is_active}{count}"
            elif exists:
                details = f" ({index.search().count()} documents)"
            else:
                details = ""
            result[module].append(f"{checkbox} {index._name}{details}")
        for app, indices in result.items():
            self.stdout.write(self.style.MIGRATE_LABEL(app))
            self.stdout.write("\n".join(indices))

    def _manage_index(self, action, indices, suffix, force, verbosity, ignore_error, **options):  # noqa
        """Manage the creation and deletion of indices."""
        action = OpensearchAction(action)

        suffix = suffix or datetime.now().strftime("%Y%m%d%H%M%S%f")

        if action == OpensearchAction.CREATE:
            known = registry.get_indices()
        else:
            known = []
            for document in [i._doc_types[0] for i in registry.get_indices()]:
                known.extend(document.get_all_indices())

        # Filter indices
        if indices:
            # Ensure every given indices exists
            known_name = [i._name for i in known]  # noqa
            unknown = set(indices) - set(known_name)
            if unknown:
                self.stderr.write(f"Unknown indices '{list(unknown)}', choices are: '{known_name}'")
                exit(1)

            # Only keep given indices
            indices = list(filter(lambda i: i._name in indices, known))  # noqa
        else:
            indices = known

        # Display expected action
        if verbosity or not force:
            if not indices:
                self.stdout.write("Nothing to do, exiting")
                exit(0)
            self.stdout.write(f"The following indices will be {action.past}:")
            for index in indices:
                index_name = index._name  # noqa
                if action == OpensearchAction.CREATE:
                    index_name += f"{Document.VERSION_NAME_SEPARATOR}{suffix}"
                self.stdout.write(f"\t- {index_name}.")  # noqa
            self.stdout.write("")

        # Ask for confirmation to continue
        if not force:  # pragma: no cover
            while True:
                p = input("Continue ? [y]es [n]o : ")
                if p.lower() in ["yes", "y"]:
                    self.stdout.write("")
                    break
                elif p.lower() in ["no", "n"]:
                    exit(1)

        pp = action.present_participle.title()
        for index in indices:
            index_name = index._name  # noqa
            if action == OpensearchAction.CREATE:
                index_name += f"{Document.VERSION_NAME_SEPARATOR}{suffix}"

            if verbosity:
                self.stdout.write(
                    f"{pp} index '{index_name}'...\r",
                    ending="",
                )  # noqa
                self.stdout.flush()
            try:
                if action == OpensearchAction.CREATE:
                    index._doc_types[0].init(suffix=suffix)
                elif action == OpensearchAction.DELETE:
                    index.delete()
            except opensearchpy.exceptions.NotFoundError:
                if verbosity or not ignore_error:
                    self.stderr.write(f"{pp} index '{index._name}'...{self.style.ERROR('Error (not found)')}")  # noqa
                if not ignore_error:
                    self.stderr.write("exiting...")
                    exit(1)
            except opensearchpy.exceptions.RequestError:
                if verbosity or not ignore_error:
                    self.stderr.write(f"{pp} index '{index._name}'... {self.style.ERROR('Error')}")  # noqa
                if not ignore_error:
                    self.stderr.write("exiting...")
                    exit(1)
            else:
                if verbosity:
                    self.stdout.write(f"{pp} index '{index_name}'... {self.style.SUCCESS('OK')}")  # noqa

    def _manage_document(
        self,
        action,
        indices,
        index_suffix,
        force,
        filters,
        excludes,
        verbosity,
        parallel,
        count,
        refresh,
        missing,
        **options,
    ):  # noqa
        """Manage the creation and deletion of indices."""
        action = OpensearchAction(action)
        known = registry.get_indices()
        filter_ = functools.reduce(operator.and_, (Q(**{k: v}) for k, v in filters)) if filters else None
        exclude = functools.reduce(operator.and_, (Q(**{k: v}) for k, v in excludes)) if excludes else None

        # Filter indices
        if indices:
            # Ensure every given indices exists
            known_name = [i._name for i in known]  # noqa
            unknown = set(indices) - set(known_name)
            if unknown:
                self.stderr.write(f"Unknown indices '{list(unknown)}', choices are: '{known_name}'")
                exit(1)

            # Only keep given indices
            indices = list(filter(lambda i: i._name in indices, known))  # noqa
        else:
            indices = known

        # Ensure every indices needed are created
        not_created = [i._name for i in indices if not i.exists()]  # noqa
        if not_created:
            self.stderr.write(f"The following indices are not created : {not_created}")
            self.stderr.write("Use 'python3 manage.py opensearch list' to list indices' state.")
            exit(1)

        # Check field, preparing to display expected actions
        s = f"The following documents will be {action.past}:"
        kwargs_list = []
        for index in indices:
            # Handle --missing
            exclude_ = exclude
            if missing and action == OpensearchAction.INDEX:
                q = Q(pk__in=[h.meta.id for h in index.search().extra(stored_fields=[]).scan()])
                exclude_ = exclude_ & q if exclude_ is not None else q

            document = index._doc_types[0]()  # noqa
            try:
                kwargs_list.append({"filter_": filter_, "exclude": exclude_, "count": count})
                qs = document.get_queryset(filter_=filter_, exclude=exclude_, count=count).count()
            except FieldError as e:
                model = index._doc_types[0].django.model.__name__  # noqa
                self.stderr.write(f"Error while filtering on '{model}' (from index '{index._name}'):\n{e}'")  # noqa
                exit(1)
            else:
                if action == OpensearchAction.MIGRATE:
                    prefix = ""
                else:
                    prefix = f" {qs}"
                s += f"\n\t- {prefix} {document.django.model.__name__}."

        # Display expected actions
        if verbosity or not force:
            self.stdout.write(s + "\n\n")

        # Ask for confirmation to continue
        if not force:  # pragma: no cover
            while True:
                p = input("Continue ? [y]es [n]o : ")
                if p.lower() in ["yes", "y"]:
                    self.stdout.write("\n")
                    break
                elif p.lower() in ["no", "n"]:
                    exit(1)

        result = "\n"
        for index, kwargs in zip(indices, kwargs_list):
            document = index._doc_types[0]()  # noqa
            model = document.django.model.__name__

            if action == OpensearchAction.MIGRATE:
                document.migrate(index_suffix)
                if verbosity >= 1:
                    result += f"{model} successfully migrated to index {document.get_index_name(suffix=index_suffix)}\n"
            else:
                qs = document.get_indexing_queryset(stdout=self.stdout._out, verbose=verbosity, action=action, **kwargs)
                success, errors = document.update(
                    qs, action, parallel=parallel, index_suffix=index_suffix, refresh=refresh, raise_on_error=False
                )

                success_str = self.style.SUCCESS(success) if success else success
                errors_str = self.style.ERROR(len(errors)) if errors else len(errors)

                if verbosity == 1:
                    result += f"{success_str} {model} successfully {action.past}, {errors_str} errors:\n"
                    reasons = defaultdict(int)
                    for e in errors:  # Count occurrence of each error
                        error = e.get(action, {"result": "unknown error"}).get("result", "unknown error")
                        reasons[error] += 1
                    for reasons, total in reasons.items():
                        result += f"    - {reasons} : {total}\n"

                if verbosity > 1:
                    result += f"{success_str} {model} successfully {action}d, {errors_str} errors:\n {errors}\n"

        if verbosity:
            self.stdout.write(result + "\n")

    def add_arguments(self, parser):
        """Add arguments to parser."""
        parser.formatter_class = argparse.RawTextHelpFormatter
        subparsers = parser.add_subparsers()

        # 'list' subcommand
        subparser = subparsers.add_parser(
            "list",
            help="Show all available indices (and their state) for the current project.",
            description="Show all available indices (and their state) for the current project.",
        )
        subparser.set_defaults(func=self.__list_index)

        # 'index' subcommand
        subparser = subparsers.add_parser(
            "index",
            help="Manage the deletion of indices.",
            description="Manage the deletion of indices.",
        )
        subparser.set_defaults(func=self._manage_index)
        subparser.add_argument(
            "action",
            type=str,
            help="Whether you want to create or delete the indices.",
            choices=[
                OpensearchAction.CREATE.value,
                OpensearchAction.DELETE.value,
            ],
        )
        subparser.add_argument(
            "--suffix",
            type=str,
            default=None,
            help="A suffix for the index name to create/delete (if you don't provide one, a timestamp will be used for creation).",
        )
        subparser.add_argument("--force", action="store_true", default=False, help="Do not ask for confirmation.")
        subparser.add_argument("--ignore-error", action="store_true", default=False, help="Do not stop on error.")
        subparser.add_argument(
            "indices",
            type=str,
            nargs="*",
            metavar="INDEX",
            help="Only manage the given indices.",
        )

        # 'document' subcommand
        subparser = subparsers.add_parser(
            "document",
            help="Manage the indexation and creation of documents.",
            description="Manage the indexation and creation of documents.",
            formatter_class=argparse.RawTextHelpFormatter,
        )
        subparser.set_defaults(func=self._manage_document)
        subparser.add_argument(
            "action",
            type=str,
            help="Whether you want to index, delete or update documents in indices. Or migrate from an index to another.",
            choices=[
                OpensearchAction.INDEX.value,
                OpensearchAction.DELETE.value,
                OpensearchAction.UPDATE.value,
                OpensearchAction.MIGRATE.value,
            ],
        )
        subparser.add_argument(
            "-f",
            "--filters",
            type=self.db_filter(parser),
            nargs="*",
            help=(
                "Filter object in the queryset. Argument must be formatted as '[lookup]=[value]', e.g. "
                "'document_date__gte=2020-05-21.\n"
                "The accepted value type are:\n"
                "  - 'None' ('[lookup]=')\n"
                "  - 'bool' ('[lookup]=True')\n"
                "  - 'float' ('[lookup]=1.12')\n"
                "  - 'int' ('[lookup]=23')\n"
                "  - 'datetime.date' ('[lookup]=2020-10-08')\n"
                "  - 'list' ('[lookup]=1,2,3,4') Value between comma ',' can be of any other accepted value type\n"
                "  - 'str' ('[lookup]=week') Value that didn't match any type above will be interpreted as a str\n"
                "The list of lookup function can be found here: "
                "https://docs.djangoproject.com/en/dev/ref/models/querysets/#field-lookups"
            ),
        )
        subparser.add_argument(
            "-e",
            "--excludes",
            type=self.db_filter(parser),
            nargs="*",
            help=(
                "Exclude objects from the queryset. Argument must be formatted as '[lookup]=[value]', see '--filters' "
                "for more information"
            ),
        )
        subparser.add_argument("--force", action="store_true", default=False, help="Do not ask for confirmation.")
        subparser.add_argument(
            "--index-suffix",
            type=str,
            default=None,
            help="The suffix for the index name (if you don't provide one, the current index will be used). Required for `migrate` subcommand.",
        )
        subparser.add_argument(
            "-i", "--indices", type=str, nargs="*", help="Only update documents on the given indices."
        )
        subparser.add_argument(
            "-c", "--count", type=int, default=None, help="Update at most COUNT objects (0 to index everything)."
        )
        subparser.add_argument(
            "-p",
            "--parallel",
            action="store_true",
            default=False,
            help="Parallelize the communication with Opensearch.",
        )
        subparser.add_argument(
            "-r",
            "--refresh",
            action="store_true",
            default=False,
            help="Make operations performed on the indices immediatly available for search.",
        )
        subparser.add_argument(
            "-m",
            "--missing",
            action="store_true",
            default=False,
            help="When used with 'index' action, only index documents not indexed yet.",
        )

        self.usage = parser.format_usage()

    def handle(self, *args, **options):
        """Run the command according to `options`."""
        if "func" not in options:  # pragma: no cover
            self.stderr.write(self.usage)
            self.stderr.write(f"manage.py opensearch: error: no subcommand provided.")
            exit(1)

        options["func"](**options)
